protocol/daemons/server/types/health_monitor.go

```diff
@@ -0,0 +1,207 @@
+// checker is lazily created using the provided function if needed. This method is synchronized. It returns an error if
+// the service was already registered.
+func (ms *healthMonitorMutableState) RegisterHealthChecker(
+	checkable types.HealthCheckable,
+	lazyHealthCheckerCreator func() *healthChecker,
+) error {
+	ms.Lock()
+	defer ms.Unlock()
+
+	// Don't register daemon services if the monitor has been disabled.
+	if ms.disabled {
+		return nil
+	}
+
+	// Don't register additional daemon services if the monitor has already been stopped.
+	// This could be a concern for short-running integration test cases, where the network
+	// stops before all daemon services have been registered.
+	if ms.stopped {
+		// If the service is stoppable, stop it. This helps us to clean up daemon services in test cases
+		// where the monitor is stopped before all daemon services have been registered.
+		if stoppable, ok := checkable.(Stoppable); ok {
+			stoppable.Stop()
+		}
+		return nil
+	}
+
+	if _, ok := ms.serviceToHealthChecker[checkable.ServiceName()]; ok {
+		return fmt.Errorf("service %v already registered", checkable.ServiceName())
+	}
+
+	ms.serviceToHealthChecker[checkable.ServiceName()] = lazyHealthCheckerCreator()
+	return nil
+}
+
+// HealthMonitor monitors the health of daemon services, which implement the HealthCheckable interface. If a
+// registered health-checkable service sustains an unhealthy state for the maximum acceptable unhealthy duration,
+// the monitor will execute a callback function.
+type HealthMonitor struct {
+	mutableState *healthMonitorMutableState
+
+	// These fields are initialized in NewHealthMonitor and are not modified after initialization.
+	logger             log.Logger
+	startupGracePeriod time.Duration
+	pollingFrequency   time.Duration
+}
+
+// NewHealthMonitor creates a new health monitor.
+func NewHealthMonitor(
+	startupGracePeriod time.Duration,
+	pollingFrequency time.Duration,
+	logger log.Logger,
+) *HealthMonitor {
+	return &HealthMonitor{
+		mutableState:       NewHealthMonitorMutableState(),
+		logger:             logger.With(cosmoslog.ModuleKey, "health-monitor"),
+		startupGracePeriod: startupGracePeriod,
+		pollingFrequency:   pollingFrequency,
+	}
+}
+
+func (hm *HealthMonitor) DisableForTesting() {
+	hm.mutableState.DisableForTesting()
+}
+
+// RegisterServiceWithCallback registers a HealthCheckable with the health monitor. If the service
+// stays unhealthy every time it is polled during the maximum acceptable unhealthy duration, the monitor will
+// execute the callback function.
+// This method is synchronized. The method returns an error if the service was already registered or the
+// monitor has already been stopped. If the monitor has been stopped, this method will proactively stop the
+// health-checkable service before returning.
+func (hm *HealthMonitor) RegisterServiceWithCallback(
+	hc types.HealthCheckable,
+	maximumAcceptableUnhealthyDuration time.Duration,
```

Comment:
similar comment as in the other PR. should we keep this consistent w the flag name? `maxDaemonUnhealthyDuration` or something similar?

Reply:
This is already addressed in the following PR, which is branched off of this branch, and will be merged into this branch before both are committed to main (in order to keep main in a releasable state.) For the sake of keeping things non-confusing and avoiding some rebasing churn, I'd like to keep this change in that PR if you don't mind!